### PR TITLE
Block cache tracing: Associate a unique id with Get and MultiGet

### PR DESCRIPTION
**Pull request by HaoyuHuang**
This PR associates a unique id with Get and MultiGet. This enables us to track how many blocks a Get/MultiGet request accesses. We can also measure the impact of row cache vs block cache. 

Test plan: make clean && COMPILE_WITH_ASAN=1 make check -j32

### CODE REVIEW

File: trace_replay/block_cache_tracer.cc [hunk 1]
```diff
@@ -266,4 +280,16 @@ Status BlockCacheTracer::WriteBlockAccess(const BlockCacheTraceRecord& record,
                                           referenced_key);
 }
 
+uint64_t BlockCacheTracer::NextGetId() {
+  if (!writer_.load(std::memory_order_relaxed)) {
+    return 0;
+  }
+  InstrumentedMutexLock lock_guard(&get_id_counter_mutex_);
+  if (get_id_counter_ == port::kMaxUint64) {
```
Comment (riversand963) [inline trace_replay/block_cache_tracer.cc]: Can you implement this with atomic variable?

Comment (HaoyuHuang) [reply trace_replay/block_cache_tracer.cc]: Sure. I removed the mutex and updated to use atomic variable.

**Comment by siying** [comment]
@HaoyuHuang clang analyze seems to fail with this PR:
```
table/block_based/block_based_table_reader.cc:3450:16: warning: Called C++ object pointer is null
          if (!get_context->SaveValue(
               ^~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```
Can you take a look?

**Comment by miasantreble** [comment]
@HaoyuHuang make check is now failing for me with name shadowing warning:
```
table/get_context.cc: In constructor ‘rocksdb::GetContext::GetContext(const rocksdb::Comparator*, const rocksdb::MergeOperator*, rocksdb::Logger*, rocksdb::Statistics*, rocksdb::GetContext::GetState, const rocksdb::Slice&, rocksdb::PinnableSlice*, bool*, rocksdb::MergeContext*, rocksdb::SequenceNumber*, rocksdb::Env*, rocksdb::SequenceNumber*, rocksdb::PinnedIteratorsManager*, rocksdb::ReadCallback*, bool*, uint64_t)’:
table/get_context.cc:47:49: error: declaration of ‘tracing_get_id’ shadows a member of 'this' [-Werror=shadow]
     bool* is_blob_index, uint64_t tracing_get_id)
```
Honestly I'm not sure what's the best practice here. Shall we rename the function to get_tracing_get_id()? Shall we skip the check for this particular type of warning?

**Comment by siying** [comment]
@miasantreble yes that will work.

**Comment by siying** [comment]
Or as long term best practice, if GetContext should be a class.

**Pull request closed without merging**